perf_bench_test.go

```diff
@@ -24,127 +24,234 @@ package main
-		images             = getImages()
-		funcs              = []string{}
+		isRunPerf          = *perfExecTime > 0
+		images             = getImages(t)
 		timeInterval       = time.Duration(time.Second.Nanoseconds() / int64(*targetReqPerSec))
 		totalRequests      = *executionTime * *targetReqPerSec
+		perfStat           = profile.NewPerfStat(*perfEvents, "perf-tmp.data", *perfExecTime, *perfInterval)
 	)
 
 	log.SetLevel(log.InfoLevel)
 
-	funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
-
 	createResultsDir()
 
-	// Pull images
-	for funcName, imageName := range images {
-		resp, _, err := funcPool.Serve(context.Background(), "plr_fnc", imageName, "record")
-
-		require.NoError(t, err, "Function returned error")
-		require.Equal(t, resp.Payload, "Hello, record_response!")
-
-		// For future use
-		// createSnapshots(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Snapshot created")
-		// createRecords(t, concurrency, vmID, imageName, isSyncOffload)
-		// log.Info("Record done")
+	if pervVMNum == 0 {
+		funcPool = NewFuncPool(!isSaveMemoryConst, servedTh, pinnedFuncNum, isTestModeConst)
 
-		funcs = append(funcs, funcName)
+		// Pull images
+		for _, imageName := range images {
+			resp, _, err := funcPool.Serve(context.Background(), "plr_fnc", imageName, "record")
+			require.NoError(t, err, "Function returned error")
+			require.Equal(t, resp.Payload, "Hello, record_response!")
+		}
 	}
 
 	// Boot VMs
 	for i := 0; i < *vmNum; i++ {
 		vmIDString := strconv.Itoa(i)
-		_, err := funcPool.AddInstance(vmIDString, images[funcs[i%len(funcs)]])
+		_, err := funcPool.AddInstance(vmIDString, images[i%len(images)])
 		require.NoError(t, err, "Function returned error")
 	}
 
-	if !*isWithCache && *isColdStart {
-		log.Info("Profile cold start")
-		dropPageCache()
-	}
+	var (
+		execTime, realRequests int64
+		vmGroup                sync.WaitGroup
+		ticker                 = time.NewTicker(timeInterval)
+		remainingRequests      = totalRequests
+	)
 
-	ticker := time.NewTicker(timeInterval)
-	var vmGroup sync.WaitGroup
-	done := make(chan bool, 1)
+	if isRunPerf {
+		err := perfStat.Run()
+		require.NoError(t, err, "Start perf stat returned error")
+	}
 
-	for totalRequests > 0 {
-		select {
-		case <-ticker.C:
-			totalRequests--
+	for remainingRequests > 0 {
+		if tickerT := <-ticker.C; !tickerT.IsZero() {
 			vmGroup.Add(1)
+			remainingRequests--
 
-			funcName := funcs[vmID%len(funcs)]
-			imageName := images[funcName]
+			imageName := images[vmID%len(images)]
 			vmIDString := strconv.Itoa(vmID)
 
-			tStart := time.Now()
-			go serveVM(t, tStart, vmIDString, imageName, &vmGroup, isSyncOffload)
+			go serveVM(t, vmIDString, imageName, &vmGroup, isSyncOffload, &execTime, &realRequests)
 
 			vmID = (vmID + 1) % *vmNum
-		case <-done:
-			ticker.Stop()
+			if vmID == 0 {
+				perfStat.SetWarmTime()
+			}
 		}
 	}
 
-	done <- true
+	ticker.Stop()
 	vmGroup.Wait()
+	perfStat.SetTearDownTime()
+
```

Comment:
the benchmark should not leave VMs hanging upon its completion. You can add a parameter `-keepAlive` but this should not be the default behavior.

Reply:
Done